File: 2021/☃️November/1119/17836.cpp
//  template.cpp
//  Copyright © 2021 도찡. All rights reserved.

#ifndef _CRT_SECURE_NO_WARNINGS
#define _CRT_SECURE_NO_WARNINGS
#endif
#define _USE_MATH_DEFINES

#include <algorithm>
#include <cstring>
#include <iostream>
#include <vector>
#include <cmath>
#include <queue>

using namespace std;
using pii = pair<int, int>;
using lint = long long;

/* [🦔] [💜] [✨] */

int R, C, T, arr[102][102];
bool visited[102][102];

const int dr[4] = {0, 1, 0, -1};
const int dc[4] = {1, 0, -1, 0};

int main() {
#ifndef ONLINE_JUDGE
#endif

    ios_base::sync_with_stdio(false); cin.tie(0); cout.tie(0);

    cin >> R >> C >> T;

    for(int r=0; r<R; r++) {
        for(int c=0; c<C; c++) {
            cin >> arr[r][c];
        }
    }

    int ans = 10001;
    queue<pair<pii, int>> qu;
    qu.push({{0,0}, 0});

    while(!qu.empty()) {
        int time = qu.front().second;
        pii current = qu.front().first;
        qu.pop();

        if(current.first==R-1 && current.second==C-1) {
            ans = min(ans, time);
            break;
        }

        if(arr[current.first][current.second] == 2) {
            ans = min(ans, time + R - current.first + C - current.second - 2);
        }

        for(int i=0; i<4; i++) {
            pii next = {current.first + dr[i], current.second + dc[i]};

            if(next.first<0 || next.first>=R || next.second<0 || next.second>=C) continue;
            if(visited[next.first][next.second] || arr[next.first][next.second] == 1) continue;

            visited[next.first][next.second] = true;
            qu.push({next, time+1});
        }
    }

    if(ans > T) cout << "Fail" << '\n';
    else cout << ans << '\n';

    return 0;
}

File: 2021/☃️November/1119/17836.swift
//  template.swift
//  Copyright © 2021 도찡. All rights reserved.

import Foundation

func readInt() -> [Int] {readLine()!.components(separatedBy: " ").map{Int(String($0))!}}
func readStr() -> [String] {readLine()!.split(separator: " ").map{String($0)}}

typealias pii = (Int, Int)
let INF = INT32_MAX
let MOD = 1e9 + 7

/*[😈][🔮][🦔]*/

struct Pos {
    var r: Int
    var c: Int
}

func main() {
    let dr = [0, 1, 0, -1]
    let dc = [1, 0, -1, 0]

    var input = readInt()
    let R = input[0]
    let C = input[1]
    let T = input[2]
    var arr = [[Int]]()
    var visited = [[Bool]](repeating: [Bool](repeating: false, count: C), count: R)

    for r in 0..<R {
        input = readInt()
        arr.append(input)
    }

    var ans = 10001
    var qu = [(pos: Pos, time: Int)]()
    var front = 0
    qu.append((Pos(r: 0, c: 0), 0))

    while qu.count > front {
        let time = qu[front].time
        let current = qu[front].pos
        front += 1

        if current.r == R-1 && current.c == C-1 {
            ans = min(ans, time)
            break
        }

        if arr[current.r][current.c] == 2 {
            ans = min(ans, time + R - current.r + C - current.c - 2)
        }

        for i in 0..<4 {
            let next = Pos(r: current.r + dr[i], c: current.c + dc[i])

            if next.r < 0 || next.r >= R || next.c < 0 || next.c >= C { continue }
            if visited[next.r][next.c] || arr[next.r][next.c] == 1 { continue }

            visited[next.r][next.c] = true
            qu.append((next, time+1))
        }
    }

    if ans > T { print("Fail") }
    else { print(ans) }
}
